pkg/apis/core/doc.go

```diff
@@ -12,6 +12,8 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-// +k8s:deepcopy-gen=package
+// +k8s:deepcopy-gen=package,register
 
-package componentconfig // import "github.com/gardener/gardener/pkg/apis/componentconfig"
+// Package core is the internal version of the API.
+// +groupName=gardener.cloud
+package core
```

Comment:
`core.gardener.cloud` group is okay.

I don't think that `ControllerInstallation` and `ControllerRegistration` should be in `core.gardener.cloud` as  **core** components. Maybe the more correct place would be in `extensions.gardener.cloud`.

 I would like to have in `core.gardener.cloud` only the absolute minimum for creating a kubernetes control plane in a K8S namespace. This being API resources responsible for `etcd`, `kube-apiserver`, `kube-controller-manager` and `kube-scheduler` (without any cloud-specific configuration. At all) reconciliation. 

Even our `Shoot` specification is too large and spanning too (for my taste) broad surface. I would hesitate of putting it as **core** resource.

Reply:
`extensions.gardener.cloud` is an API group installed via CRD to all seeds whereas `core.gardener.cloud` is only available in the Garden cluster and installed via APIService.

I'm not sure why you haven't responded on the GEP back then when it was in review with these comments. I don't see how we can continue without the resources. Theoretically, we could move them to another API group group managed by Gardener API server in the Garden cluster, but even then I don't see benefits. IMHO these resources belong to the core - Gardener will need the extension controllers to incorporate vendor specifics, and the registration part is core to the system.

Reply:
I tend to agree with @mvladev in the sense that packaging decisions are sometimes orthogonal to how the actual running system works. That is, `core` doesn't necessarily means self-suficient. 
  
That said, from @rfranzke  explanation it seams like a more natural package structure would be separating elements from the `gardener` (a.k.a `core`) , `seed`, and `shoot` cluster. For example, the extensions are installed in seeds, so they maybe should belong to a `seed` package.

Reply:
Thanks for the feedback! However, I'm not sure whether I get your points. Gardener extends the API of an existing K8s cluster by managing a few resources that are *core* to the system: `CloudProfile`s, `Seed`s, `Shoot`s, etc. Today all of the vendor specific code is in-tree, but as we are extracting these things into external controllers there will be the *need* to register and install them. Consequently, IMHO the new `Controller{Registration,Installation}` resources are *core* to the system as well.
I don't see any benefit in moving them into other API groups - this comes with a lot of overhead wrt API, client generation, registry, etc.
I get the point why K8s is doing that with its, 20, 30, 40, or even more resources, but we are far from that.

Reply:
Thanks for the explanation.  My comment was based on the impression I got from the description in the extensibility document that these controllers were part of the seed cluster's control plane and therefore it also made sense to consider the registry part of a `seed` not the `core`. 

In the proposal, the controllers are registered in the garden cluster and propagated (deployed) to the seed. 

I was expecting the seeds to become more autonomous under the new architecture and make the garden cluster mostly responsible for launching the seeds. The registry and management of controllers could be done in the seed. The only piece I'm not sure how to fit are the operators that manage the life-cycle of controllers. I was expecting it to be generic, because I don't see the point of having this logic part of the controller.

I must recognize that such change, splitting the api between the garden and the seed, would make the operation more complex. However the resulting operation model would be more flexible and allow, for example, that in a multi-tenant environment each seed could be managed independently. 

BTW, in my previous comment I somehow used `package` when referring to `group`, because in my mind both should be aligned, but you manage to understand it. Sorry for the cognitive overhead I introduced. :-)

Reply:
@rfranzke Thanks for that clarification.  My comment was more about the way the operators will be deployed in the seed cluster.  To put his comment in context, let me cite the paragraphs below from the [extensibility proposal](https://github.com/gardener/gardener/blob/master/docs/proposals/01-extensibility.md#registration-of-external-controllers-at-gardener):

>When a Shoot shall be reconciled Gardener can identify based on the referenced Seed and the content of the Shoot specification which controllers are needed in the respective Seed cluster. It will demand the operators in the Garden cluster to deploy the controllers they are responsible for to a specific Seed. This kind of communication happens via CRDs as well:

> The default scenario is that every controller is gets deployed by a dedicated operator that knows how to handle its lifecycle operations like deployment, update, upgrade, deletion. This operator watches ControllerInstallation resources and reacts on those it is responsible for (that it has created earlier). Gardener is responsible for writing the .spec field, the operator is responsible for providing information in the .status indicating whether the controller was successfully deployed and is ready to be used. Gardener will be also able to ask for deletion of controllers from Seeds when they are not needed there anymore by deleting the corresponding ControllerInstallation object.

> For most cases the controller deployment is very simple (just deploying it into the seed with some static configuration). In these cases it would produce unnecessary effort to ask for providing another component (the operator) that deploys the controller. To simplify this situation Gardener will be able to react on ControllerInstallations specifying .spec.registration.deployment.type=helm

All the above discussion about how the operators should be installed makes me think this should be part of the seed cluster control plane, not the gardener clusters control plane, and therefore part of a `seed` api group. 

On the contrary, I understand you consider this is part of the `core` api because this is carried by the gardener cluster:
> Theoretically, we could move them to another API group group managed by Gardener API server in the Garden cluster, but even then I don't see benefits. IMHO these resources belong to the core - Gardener will need the extension controllers to incorporate vendor specifics, and the registration part is core to the system.

In summary. Where the api belongs is a matter of who's responsibility it is to manage the controllers.  think it should be the seed's responsibility because it gives more flexibility. But I also understand this is a significant change in the architecture, but it worth taking into consideration.